### DISCUSSION TITLE
How do I use `ScrollArea::show_viewport` with a TextEdit::multiline?

### DISCUSSION THREAD

**Discussion by zkldi** [emilk/egui#2902] [Q&A]
For context, I have a potentially extremely large `TextEdit::multiline` rendering inside a scroll area. This thing may span 100 thousand newlines worth of content.

Rendering this "normally" causes pretty huge performance issues (<1fps), so I was looking into `show_viewport` to optimise the painting, however, I don't really understand what I'm doing :sweat_smile:.

```rs
				let te = TextEdit::multiline(str).desired_width(f32::INFINITY);

				let te = te.code_editor();
				let te = te.interactive(false);

				ScrollArea::vertical()
					.id_source("text_editor_todo")
					.show_viewport(ui, |ui, rect| {
						ui.set_clip_rect(rect);

						te.show(ui);
					});
```
Here's what I've got and it doesn't really work - not only is it not performance, but `set_clip_rect` is breaking the actual view. What do I do with the `|ui, rect|` closure to improve performance here?

**Comment by hacknus** [comment]
Hi, I'm also looking into this (see #2906).
I don't think `ui.set_clip_rect()` is the way to go, because then we still create the full sized text to be shown which uses resources. Instead, I think we need to look at this example here about [large scrolling fields](https://github.com/emilk/egui/blob/8ce0e1c5206780e76234842b94ceb0edf5bb8b75/crates/egui_demo_lib/src/demo/scrolling.rs).

**Comment by hacknus** [comment]
I attached an example in #2906 that worked for me.
